### PR TITLE
feat: implement CasePrimitive for JIT-compatible lax.cond/lax.switch

### PR DESCRIPTION
**Pull request by tsumme1**
## Summary

Implement `CasePrimitive` for the MPS PJRT plugin, enabling `lax.cond` and `lax.switch` to work correctly under `jax.jit` with `mx::compile`. Also removes dead compile-probe infrastructure and restores a critical MLX CompilerCache patch.

Refs #98

## Problem

`mx::compile` caches intermediate computation results and does not re-evaluate them when the outer compiled function replays with new inputs. For `stablehlo.case`, the branch index is typically computed by an intermediate `clamp` operation:

```
%c = constant 0
%c_0 = constant 1
%0 = clamp(%c, %arg0, %c_0)  ← frozen by mx::compile
%1 = case(%0)                 ← receives stale index on replay
```

This causes the case primitive to always select the same branch regardless of the runtime index value.

## Solution

A three-part approach:

1. **Index op inspection** — inspect the immediate defining op of the case op's index operand. For `clamp(const, argN, const)` patterns (lax.switch), extract the root operand and clamp bounds, then build a `compiledIndexFn` that re-applies the clamp at runtime from the raw function argument. For `convert(bool→int32)` patterns (lax.cond), look through one level to find the source argument.

2. **On-demand operand addition** — when the inspection finds a root operand not already in the external values (happens in nested cases), add it from the values map.

3. **Targeted dependency collection** — collect only the external values actually referenced by branch regions (via `CollectExternalValues`) plus the root argument needed for index reconstruction. This keeps the CasePrimitive's input signature proportional to what the branches actually use, rather than scaling with the enclosing function's arity.

## Architecture

- CPU-stream custom primitive (opaque to `mx::compile`) dispatches compiled branch functions at runtime
- Matches the pattern established by `WhileLoopPrimitive`
- Handles: `clamp` patterns (lax.switch), `convert` patterns (lax.cond bool→int32), direct function arguments, constant indices

### Dead code removal (`handler_utils.h`, `control_flow.cc`, `mlx_executable.cc`)
- Removed `CompileIncompatibleError` exception class — no longer needed since `CasePrimitive` and `WhileLoopPrimitive` handle all compile-path control flow
- Removed `allow_while_primitive` flag from `ExecContext` — was always true when `inside_compile` is true, making it redundant
- Removed two unreachable `throw` blocks in `HandleWhile` and `HandleCase`
- Simplified compile-path condition from `ctx.inside_compile && ctx.allow_while_primitive` to just `ctx.inside_compile`

### CompilerCache teardown fix (`third_party/mlx/patches/07-...`)
- Restored patch `07-compiler-cache-detach-on-destroy.patch` that was prematurely dropped in PR #148
- The public `compile()` API prevents UAF on the main thread, but the thread-local `CompilerCache` destructor during `pthread_exit` still triggers re-entrant `compile_erase()` calls that hit freed hash-map nodes
- The patch moves the cache map before destroying it in both `clear()` and `~CompilerCache()`, making re-entrant erases no-ops
- Fixes SIGTRAP in `WhileLoopPrimitive::~WhileLoopPrimitive()` during multi-threaded test teardown

## Defensive guards

- `needsClamp` only set when both clamp bounds are proven-constant and parseable
- Unrecognized index defining ops (not clamp, convert, or constant) rejected with `MPS_LOG_ERROR` + `return false` instead of silently freezing
- `stablehlo::ConstantOp` indices explicitly exempted from the rejection guard (safe to freeze)

## Test Results

All case primitive tests pass, comparing MPS output against CPU reference:

| Category | Tests | Status |
|----------|-------|--------|
| `lax.cond` (true/false, scalar/array) | 4 | ✅ |
| `lax.switch` (2/3/5 branches) | 3 | ✅ |
| Out-of-bounds clamping | 2 | ✅ |
| Multiple return values | 1 | ✅ |
| 2D array inputs | 1 | ✅ |
| Many branches (5) | 1 | ✅ |
| Nested switch-in-switch | 1 | ✅ |
| Cond inside while loop | 1 | ✅ |
| Switch inside while loop | 1 | ✅ |
| Compiled replay (10 calls, varying inputs) | 1 | ✅ |

## Files Changed

- **`src/pjrt_plugin/ops/control_flow.cc`** — CasePrimitive class + HandleCase; dead code removal
- **`src/pjrt_plugin/ops/handler_utils.h`** — Removed `CompileIncompatibleError` and `allow_while_primitive`
- **`src/pjrt_plugin/mlx_executable.cc`** — Simplified exception handling
- **`third_party/mlx/patches/07-compiler-cache-detach-on-destroy.patch`** — Restored CompilerCache teardown fix

### CODE REVIEW

File: src/pjrt_plugin/ops/control_flow.cc [hunk 1]
```diff
@@ -542,8 +636,256 @@ bool HandleCase(mlir::Operation* op, ValueMap& values, std::vector<mlx::core::ar
...
+                        extArrays.push_back(valIt->second);
+                        keyToPos[operandKey] = indexArgPos;
+                    }
+                }
+                if (indexArgPos != SIZE_MAX) {
+                    // Extract clamp bounds from the MLIR constants.
+                    // Only enable clamp reconstruction when both bounds are
+                    // proven constant — otherwise fall through to the direct
+                    // argument path to avoid silently using default bounds.
+                    bool parsedMin = false;
```
[the remark this answers is not in the document]
Comment (tsumme1) [reply src/pjrt_plugin/ops/control_flow.cc:724]: Addressed — when a ClampOp is identified and the operand is resolved but either bound fails to parse as a constant, we now return false with a clear error instead of silently dropping the clamp and passing through the raw operand. This closes the gap where a clamp with non-constant bounds would produce an unguarded index that could select out-of-range branches.
